src/foam/nanos/auth/AuthenticatedAgentJunctionDAO.java

```diff
@@ -0,0 +1,164 @@
+/**
+ * @license
+ * Copyright 2018 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
+
+import foam.core.FObject;
+import foam.core.X;
+import foam.dao.DAO;
+import foam.dao.ProxyDAO;
+import foam.dao.Sink;
+import foam.mlang.order.Comparator;
+import foam.mlang.predicate.Predicate;
+import foam.util.SafetyUtil;
+
+import static foam.mlang.MLang.AND;
+import static foam.mlang.MLang.EQ;
+import static foam.mlang.MLang.OR;
+
+/**
+ * Authenticated AgentJunctionDAO
+ * Restrict users from creating a junction object where they are the source user (Agent)
+ * Restrict users from finding and selecting junction objects they are not referenced to. (Agent or Entity)
+ * Restrict users from removing junction objects where they aren't the target user or source user. (Agent or Entity)
+ */
+public class AuthenticatedAgentJunctionDAO
+  extends ProxyDAO
+{
+
+  public final static String GLOBAL_AGENT_JUNCTION_READ   = "agentJunction.read.*";
+  public final static String GLOBAL_AGENT_JUNCTION_UPDATE = "agentJunction.update.*";
+  public final static String GLOBAL_AGENT_JUNCTION_DELETE = "agentJunction.delete.*";
+
+  public AuthenticatedAgentJunctionDAO(X x, DAO delegate) {
+    super(x, delegate);
+  }
+
+  /**
+   * Allows put if user or agent is target of the junction object.
+   * If agent exists within context, condition checks on agent user.
+   */
+  @Override
+  public FObject put_(X x, FObject obj) {
```

Comment:
This is missing the code restricts what the group can be set to. If this were merged as is, anyone could give themselves super-admin privileges.

Reply:
As mentioned before I was going to address this in a later PR and account for the changes in the test but I've went ahead and added it in https://github.com/foam-framework/foam2/pull/1637/commits/3ecb2f98c65d21f438b880540237487b5a4dd3b3.